### ISSUE TITLE
Latest macOS Fiji launcher issue

### ISSUE THREAD

**Issue by hyungsongnam** [fiji/fiji#271]
Hello,
On my Mac, the latest update modified the Fiji launcher, and now it seems the update altered the launching from the Dock.
Please see attached movie for this behavior.
Thank you.
Sincerely,
Hyung-song Nam, PhD

[Screen Recording 2020-11-04 at 10.42.01 PM 720p.mov.zip](https://github.com/fiji/fiji/files/5492222/Screen.Recording.2020-11-04.at.10.42.01.PM.720p.mov.zip)

**Comment by hyungsongnam** [comment]
Upon further tinkering, I've found that this only occurs with the No-JRE version. The macOS Fiji with the JRE included does not behave the same way, it works as expected. It also allows to launch in low resolution mode. So, for now, this is the fix I was looking for.
Thank you.
Sincerely,
Hyung-song Nam, PhD

**Comment by ctrueden** [comment]
Glad you found a workaround, @hyungsongnam!

**Comment by hyungsongnam** [comment]
Thank you.
Please see attached image:

![image](https://user-images.githubusercontent.com/36607862/98448992-4ca2ae80-20ed-11eb-99f7-db2dc02eee0b.png)

**Comment by hyungsongnam** [comment]
Something is off with the No-JRE version in Windows as well.
The most recent downloaded Fiji No-JRE version does not launch at all.
The JRE included version does launch, but I get this error:

![image](https://user-images.githubusercontent.com/36607862/98449023-883d7880-20ed-11eb-88cc-1145f134d17a.png)

Thank you.

**Comment by hyungsongnam** [comment]
Hello,
I tested the updated Fiji (4 pm MST, November 9, 2020).
The updated version of No-JRE Fiji can find the system Java.
The launch in low resolution mode works as well.
So, problems fixed.
I will try on the Windows 10 machine at work tomorrow.
Thank you.
Hyung-song Nam, PhD

**Comment by ctrueden** [comment]
@hyungsongnam Thanks for these details. We recently rolled out a new version of the launcher, but in response to problems, rolled it back. I'm going to reopen this issue until we can verify that the problems you describe here are fixed in the _new_ version of the launcher, which we will eventually need to roll out again—after the known rough edges are solved.

**Comment by hyungsongnam** [comment]
Hello,
Just to inform you what I see here after a latest update, if I launch the Fiji from command line, now I see

<img width="567" alt="image" src="https://user-images.githubusercontent.com/36607862/99599860-72fffe00-29b9-11eb-976d-89e0b4ded291.png">

I'm not sure why the error occurs. The Fiji launches ok despite this, and runs from the system installed Java. I'm running the No-JRE Fiji with Oracle JDK 8 on macOS Big Sur.

Thank you.
Hyung-song Nam, PhD

**Comment by hyungsongnam** [comment]
Hello,
I would like to add another launcher related behavior noticed with Fiji on macOS Big Sur.
The background is, I use an ImageJ1 script that calls the parallel gzip program "pigz" using "exec."
If I launch Fiji on my Big Sur mac by double clicking the program icon in Applications (with or without included Java), the "exec" in the script will not run and the script will end.
However, if I launch Fiji from command line by going to Applications/Fiji/Contents/MacOS and opening ImageJ-macosx, the "exec" command will run.
If I also launch Fiji by double-clicking on the ImageJ-macosx binary in Finder, that will also run the "exec" command.
So, I guess this is because I gave Terminal permission to access Desktop folders, where the image file is located, whereas Fiji doesn't have permission to access the Desktop (a security feature on Big Sur?).
I don't mind opening Fiji from command line, but is there a way to address this such that Fiji can be opened by double-clicking the icon and still have the "exec" command run?
I tried adding Fiji, ImageJ-macosx, java, and javac to System Preferences/Security & Privacy/Privacy/Full Disk Access, but none of these granted Fiji access to the folder on the Desktop.
Thank you.
Sincerely,
Hyung-song Nam, PhD

**Comment by hyungsongnam** [comment]
Hello,
I would like to report a couple of different issues on a different system but still relating to the Fiji launcher.
Please see the attached screen capture video.
https://github.com/fiji/fiji/assets/36607862/84cafb9e-5c9b-45cb-9c77-044d19655299

The system is an Apple Silicon M2 MacBook Pro with Oracle Java 8 version 381 running the No_jre Fiji.
Two issues:
(1) After launching, Fiji shows as ImageJ on the upper left corner.
(2) The Fiji dock icon (created by dragging the /Applications/Fiji to the dock), disappears after quitting the app and changes to genetic java icon. Clicking this java icon thereafter does not launch Fiji but opens java in a terminal window.
Thank you.
Sincerely,
Hyung-song Nam, PhD

**Comment by ctrueden** [comment]
@hyungsongnam One thing I notice in your video that is odd is from the Properties window: the `java.home` value begins with `/Library/Internet Plug-Ins/JavaApplet`. I'm surprised this still folder is still present on an M2 MacBook; I thought that the Java applet browser plugin was long since retired and Apple even uninstalled it from existing systems IIRC. In any case, the Launcher is definitely picking up the wrong version of Java.

One consideration is whether you are launching Fiji in x86_64/Rosetta compatibility mode, or native arm64 mode? We don't have a native arm64 bundle of Fiji yet, although that is planned. Further reading:

* https://imagej.net/software/fiji/downloads#system-requirements
* https://forum.image.sc/t/fiji-not-opening-on-m1-macbook-pro-2020/57518/9
* https://forum.image.sc/t/fiji-for-apple-silicon-macs/80794/9
* https://github.com/ij-plugins/ijp-imagej-launcher#installing-fiji-on-mac-os-x-arm64

> I see the same with aarch64 Zulu Java 8 and no_jre Fiji, but not with the macOS-specific Fiji with bundled x86_64 Zulu Java 8.

Make sure, when you are testing this, that those versions of Java are indeed what is being used. You can force Fiji to use a specific Java using the argument `--java-home /path/to/a-jdk`, where `/path/to/a-jdk` is the path to the Java you want to be used.

**Comment by hyungsongnam** [comment]
@ctrueden 
Thank you for the reply.
I looked into the points above.
(1) The /Library/Internet Plug-Ins/JavaApplet folder does exist, and I think the Oracle Java 8 JDK installer put it there.

<img width="1728" alt="Screenshot 2023-07-27 at 3 10 20 PM" src="https://github.com/fiji/fiji/assets/36607862/62e80095-2615-4ec6-947e-19b1357f68bc">

(2) I tried the below but got an error.

<img width="1728" alt="Screenshot 2023-07-27 at 3 16 06 PM" src="https://github.com/fiji/fiji/assets/36607862/8ef72327-3ab9-4c2e-bf07-bbafc616d790">

(2) I also tried it differently, but did not seem to have changed the Java home folder.

<img width="1728" alt="Screenshot 2023-07-27 at 3 19 26 PM" src="https://github.com/fiji/fiji/assets/36607862/4c91ed90-3596-4421-bf02-874e6238856e">

Bottom line, I think I will just wait a little longer for the arm64 bundle of Fiji. Thank you for the work on Fiji.
SIncerely,
Hyung-song Nam, PhD

**Comment by ctrueden** [comment]
Based on one of your error messages, `have 'arm64', need 'x86_64'`, I suspect you are still using the `ImageJ-linux64` launcher bundled with fiji-nojre, which is not built for arm64. If you [download the latest version of the launcher](https://imagej.net/learn/launcher#download) and rename it to `ImageJ-linux64` and then `chmod +x ImageJ-linux64` on it to make it executable, it might then successfully link up with your installed arm64 Java installation. By the way, I suggest not to use the Oracle JDK, because it is not open source—instead, I suggest using the [Zulu 8 JDK+FX](https://www.azul.com/downloads/?version=java-8-lts&os=macos&architecture=arm-64-bit&package=jdk-fx#zulu).

But I know it's a pain right now to get it working, so waiting for an official arm64 bundle is totally reasonable.

**Comment by hyungsongnam** [comment]
Thank you.

I switched the Java 8 on the M2 machine to Zulu JDK+FX.

Now the Java home directory is correct. So maybe that had something to do with the Oracle Java install.

<img width="1728" alt="Screenshot 2023-07-27 at 4 17 22 PM" src="https://github.com/fiji/fiji/assets/36607862/686bed02-5d74-42b0-b223-b285d42adb34">

The Fiji install right now does work.

However, as mentioned above, the app name on the menu bar still shows as "ImageJ" and the dock icon still says "java," if I use a separate Java install and the no_jre Fiji app.

Also, as for the beta launcher, do you mean the ImageJ-macos instead of Imagej-linux? I'm testing this on a macOS machine. In any case, I downloaded both and put them in /Applications/Fiji and /Applications/Fiji/Contents/MacOS after changing the file name and the permissions. That did not change the issue I'm having with the Fiji dock icon (showing as "java" and disappearing when I quit the app).

Thank you.
Sincerely,
Hyung-song Nam, PhD

**Comment by ctrueden** [comment]
> for the beta launcher, do you mean the ImageJ-macos instead of Imagej-linux?

Doh, yes, I meant `ImageJ-macosx`, replacing the one in` Contents/MacOS` and ensuring it is executable with `chmod +x`. The new builds of the Launcher for macOS are "Universal 2" binaries including both x86_64 and arm64 architectures. If I recall correctly, you can use `otool -L ImageJ-macosx` to confirm its supported architectures with any given executable.

> the app name on the menu bar still shows as "ImageJ" and the dock icon still says "java" and it disappears when I quit the app.

Sorry, I'm not sure why that is happening. I do know that it is [this code](https://github.com/imagej/imagej-legacy/blob/imagej-legacy-1.2.1/src/main/java/net/imagej/legacy/IJ1Helper.java#L210-L215) that changes the Dock icon, by calling (via reflection) `com.apple.eawt.Application.getApplication().setDockIconImage(ImageIO.read(iconURL))`, with `iconURL` referencing the `Fiji.app/images/icon.png` file.

**Comment by hyungsongnam** [comment]
Thanks for the reply.

Just one more comment, it looks like the link above doesn't seem to download the Universal binary of the Fiji launcher.

Clicking the link downloaded a file with this file name:

<img width="1728" alt="Screenshot 2023-07-27 at 9 48 33 PM" src="https://github.com/fiji/fiji/assets/36607862/2e3c0487-29f4-4f6e-a521-9e33a5eb83d8">

Checking whether this binary was Universal showed it was only for x86_64:

<img width="1728" alt="Screenshot 2023-07-27 at 9 48 56 PM" src="https://github.com/fiji/fiji/assets/36607862/ce1c8213-3279-4765-bc82-1a21d86d80aa">

Also, the otool -L output:

<img width="1728" alt="Screenshot 2023-07-27 at 9 49 08 PM" src="https://github.com/fiji/fiji/assets/36607862/bf324b27-367d-44bc-8885-b975aed8d944">

Thank you for the explanation for the issue I'm experiencing. I'm not familiar with the Fiji code, so I can't be helpful in addressing it. However, that is what I noticed on the M2 machine.

Sincerely,
Hyung-song Nam, PhD

**Comment by ctrueden** [comment]
Yep, you're right, it's `file` to check the architectures, not `otool -L`. Sorry for misremembering that.

And about the download not being Universal 2: that's very unfortunate, and thank you very much for raising that issue.

Checking [the directory of artifacts](https://maven.scijava.org/content/repositories/snapshots/net/imagej/imagej-launcher/6.0.3-SNAPSHOT/), it is clear that the macOS version of the native launcher executable has not been built since Tue Nov 17 14:01:35 CST 2020, which is before the Universal 2 build change was made. So while you can clone the [imagej-launcher repository](https://github.com/imagej/imagej-launcher) and build it yourself, the CI has not been doing it. Why not? Because we never updated [the GitHub Actions CI workflow](https://github.com/imagej/imagej-launcher/blob/47befcbde8ce3ce114b590764a86caa6c2b4e654/.github/workflows/build-main.yml) to do a multi-platform build. :disappointed: I will try to fix this soon, as part of the planned update to Java 11.

**Comment by hyungsongnam** [comment]
Thank you. I'm glad to be helpful. I look forward to the Java 11 update.
Sincerely,
Hyung-song Nam, PhD

**Comment by hinerm** [comment]
I went through and tested the following issues using the latest [Jaunch-based launcher](https://github.com/apposed/jaunch), which is currently distributed on the [SciJava Ops update site](https://ops.scijava.org/en/1.0.0/Installation.html#fiji). I used an Arm64 Mac for these tests.

* "keep in dock”/dragging from Applications both work for pinning Fiji in the dock (and it keeps the Fiji icon)
* Launching from the dock icon works fine without falling back to a plain Java command
* Launching with Oracle Java (21) works
* An ImageJ1 script that calls exec runs appropriately, whether Fiji was launched by double-clicked or from terminal (but may only be able to interact with certain locations based on Mac security)
* Application menu shows “Fiji”, not “ImageJ”

I think this covers everything in this thread. If new things come up with the new launcher, please let us know!

**Comment by ctrueden** [comment]
> maybe it will be permanent going forward.

Yes, unfortunately, macOS's increasingly strict security model makes it infeasible for Fiji to store all the JAR files and other components inside of the .app itself without [triggering Gatekeeper telling the user to throw the app away](https://github.com/apposed/jaunch/blob/1.0.2/doc/MACOS.md#code-signing). So a folder it is!
